sweagent/agent/models.py

```diff
@@ -425,6 +428,157 @@ def query(self, history: list[dict[str, str]]) -> str:
         response = "\n".join([x.text for x in response.content])
         return response
 
+class BedrockModel(BaseModel):
+    MODELS = {
+        "anthropic.claude-instant-v1": {
+            "max_context": 100_000,
+            "cost_per_input_token": 8e-07,
+            "cost_per_output_token": 2.4e-06,
+        },
+        "anthropic.claude-v2": {
+            "max_context": 100_000,
+            "cost_per_input_token": 8e-06,
+            "cost_per_output_token": 2.4e-05,
+        },
+        "anthropic.claude-v2:1": {
+            "max_context": 100_000,
+            "cost_per_input_token": 8e-06,
+            "cost_per_output_token": 2.4e-05,
+        },
+        "anthropic.claude-3-sonnet-20240229-v1:0": {
+            "max_context": 200_000,
+            "max_tokens": 4096,  # Max tokens to generate for Claude 3 models
+            "cost_per_input_token": 3e-06,
+            "cost_per_output_token": 1.5e-05,
+        },
+        "anthropic.claude-3-haiku-20240307-v1:0": {
+            "max_context": 200_000,
+            "max_tokens": 4096,
+            "cost_per_input_token": 2.5e-07,
+            "cost_per_output_token": 1.25e-06,
+        },
+    }
+
+    def __init__(self, args: ModelArguments, commands: list[Command]):
+        super().__init__(args, commands)
+
+        model_provider = self.api_model.split('.')[0]
+        if model_provider == "anthropic":
+            # Note: this assumes AWS credentials are already configured.
+            # https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html
+            self.api = AnthropicBedrock()
+        else:
+            raise NotImplementedError
+
+    def history_to_messages(
```

Comment:
How specific is this code and that of `query` to this class? Can't we just inherit most of this from the anthropic model? Really should cut down on duplicity.

Reply:
Thanks for reviewing this, @klieret. I was more worried about future proofing the changes than about code duplication. Even though this PR only adds support for Claude models, the idea is to prepare the class for adding other models and model providers.

```bash
def history_to_messages(...) -> list[dict[str, str]]:
    (...)
    model_provider = self.api_model.split('.')[0]
    if model_provider == "anthropic":
        # Covered by this PR
        # Adapted from the AnthropicModel implementation
    elif model_provider == "cohere":
        # Not supported
    elif model_provider == "meta":
        # Idem
    (...)
    else
        raise NoImplementedError(f"{self.api_model} is not supported!")
```

For `anthropic` models, the code is the same because you can just replace `Anthropic` with `AnthropicBedrock`, but this may not always be the case. Plus, models have different IDs e.g. `claude-3-sonnet-20240229` (1P) vs `anthropic.claude-3-sonnet-20240229-v1:0` (Bedrock) and can different lifecycles.

Reply:
I understand, inheritance might indeed not be the best way then. Still, couldn't we extract some of the shared code for anthropic-style interfaces into e.g., a function that is shared between `Anthropic` and `AnthropicBedrock`?

Reply:
But hoping to get this merged this week :)

Reply:
The only difference is the model name (*see images*), so it's feasible. Just have to split the `api_model` further. I'll do that in a sec.

**Anthropic**

![image](https://github.com/princeton-nlp/SWE-agent/assets/7282984/dcfe39f2-3d45-46d9-8f62-aaaba4faff6f)

**Bedrock**

![image](https://github.com/princeton-nlp/SWE-agent/assets/7282984/eb8d6b65-4652-464b-8449-e8c825adead7)

Reply:
@klieret Changes implemented ✅ 1/ refactored `AnthropicModel` and `BedrockModel` to avoid code duplication (created methods `anthropic_history_to_messages` and `anthropic_query`), 2/ added more informative error messages (incl. model provider validation) to `BedrockModel` and 3/ added [Claude 3 Opus](https://aws.amazon.com/blogs/aws/anthropics-claude-3-opus-model-on-amazon-bedrock/)